### PR TITLE
CMAKE CUDA fix: Unknown option '-Wall'

### LINKED ISSUE #5542: CMAKE error nvcc fatal   : Unknown option '-Wall'
On Ubuntu 22.04, using CMAKE and -DLLAMA_CUBLAS=ON I'm getting the following error:

```
Cannot get compiler information:
	Compiler exited with error code 1: /usr/local/cuda/bin/nvcc -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DGGML_CUDA_PEER_MAX_BATCH_SIZE=128 -DGGML_USE_CUBLAS -DK_QUANTS_PER_ITERATION=2 -D_GNU_SOURCE -D_XOPEN_SOURCE=600 -isystem=/usr/local/cuda/include -O3 -DNDEBUG -std=c++11 --generate-code=arch=compute_52,code=[compute_52,sm_52] --generate-code=arch=compute_61,code=[compute_61,sm_61] --generate-code=arch=compute_70,code=[compute_70,sm_70] -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -use_fast_math -Wno-pedantic -Xcompiler "-Wno-array-bounds -Wno-format-truncation -Wextra-semi -march=native" --dryrun /tmp/compiler-file17262593416218535232.cu
	nvcc fatal   : Unknown option '-Wall'
```

The problem can be solved by replacing the following lines in the CMakeLists.txt:

```
set(CUDA_FLAGS ${CXX_FLAGS} -use_fast_math)
if (NOT MSVC)
    list(APPEND CUDA_FLAGS -Wno-pedantic)
endif()
```

to:

```
set(CUDA_FLAGS -use_fast_math)
if (NOT MSVC)
    list(APPEND CUDA_FLAGS --compiler-options=-Wno-pedantic)
endif()
```

Maybe someone have better idea how to include "${CXX_FLAGS}" for CUDA_FLAGS?

### PR DESCRIPTION
**Pull request by clibdev**
Closes #5542

### CODE REVIEW

**Comment by phymbert** [comment]
I was also facing this issue with CLion CMake import, thanks a lot @clibdev

File: CMakeLists.txt [hunk 1]
```diff
@@ -767,9 +767,10 @@ endif()
 set(CUDA_CXX_FLAGS "")
 
 if (LLAMA_CUBLAS)
-    set(CUDA_FLAGS ${CXX_FLAGS} -use_fast_math)
+    set(CUDA_FLAGS -use_fast_math)
+    list(APPEND CUDA_CXX_FLAGS ${CXX_FLAGS})
```
Comment (slaren) [inline CMakeLists.txt:771]: I think this line can be removed entirely, the `CXX_FLAGS` is used for warnings, and that is already being passed to `nvcc` in the next block.
```suggestion
```

Comment (clibdev) [reply CMakeLists.txt:782]: Inside `build` directory: `cat CMakeFiles/ggml.dir/flags.make | grep Xcompiler`

Before removing line:

```
CUDA_FLAGS = -O3 -DNDEBUG -std=c++11 "--generate-code=arch=compute_52,code=[compute_52,sm_52]" "--generate-code=arch=compute_61,code=[compute_61,sm_61]" "--generate-code=arch=compute_70,code=[compute_70,sm_70]" -use_fast_math -Xcompiler "-Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -Wno-pedantic -Wno-array-bounds -Wno-format-truncation -Wextra-semi -march=native"
```

After removing line:

```
CUDA_FLAGS = -O3 -DNDEBUG -std=c++11 "--generate-code=arch=compute_52,code=[compute_52,sm_52]" "--generate-code=arch=compute_61,code=[compute_61,sm_61]" "--generate-code=arch=compute_70,code=[compute_70,sm_70]" -use_fast_math -Xcompiler "-Wno-pedantic -Wno-array-bounds -Wno-format-truncation -Wextra-semi -march=native"
```

Can you double-check?

Comment (slaren) [reply CMakeLists.txt:782]: Yes, it is still wrong with that change alone. There are a few inconsistencies in the way the CUDA host compiler flags are being set.

**Review by slaren — APPROVED** [review]
I think we can merge this to solve the immediate problem, but the way `CXX_FLAGS` and the function `get_flags` is being used looks completely wrong to me. `get_flags` is only used with `LLAMA_ALL_WARNINGS`, yet it also contains optimization flags for the Intel compiler. `get_flags` is also used with CUDA to set the host compiler flags, and it actually checks what is the host compiler instead of assuming that it is the same than for the rest of the C++ code, which is great, but there are still flags from `CXX_FLAGS` that should also be passed to the host compiler. Adding `CXX_FLAGS` to `CUDA_CXX_FLAGS` is not correct either, because `CXX_FLAGS` also includes the flags from `get_flags`, but checked for the wrong compiler.

**Comment by ggerganov** [comment]
@slaren Thanks for the review

> get_flags is only used with LLAMA_ALL_WARNINGS, yet it also contains optimization flags for the Intel compiler

@abhilash1910 Could you take a look into this - it does look incorrect to add Intel compiler flags here

Will wait for @cebtenzzre to review `get_flags` usage and see if we can improve

**Comment by cebtenzzre** [comment]
> `get_flags` is only used with `LLAMA_ALL_WARNINGS`, yet it also contains optimization flags for the Intel compiler.

This was added by @ngxson in #5068, I didn't review that PR because the title and commit messages only mentioned Docker.

> Adding `CXX_FLAGS` to `CUDA_CXX_FLAGS` is not correct either, because `CXX_FLAGS` also includes the flags from `get_flags`, but checked for the wrong compiler.

The idea is that CXX_FLAGS contains generic flags, and GF_CXX_FLAGS contains the compiler-specific flags - they must both be combined to get the full list of flags. For the host this is done in the add_compile_options step, for CUDA this is currently not done consistently.

**Comment by cebtenzzre** [comment]
I propose this solution: #5570

**Comment by slaren** [comment]
I don't think this works correctly with GNU systems, `make` prints this:

```
expr: syntax error: unexpected argument ‘070100’
expr: syntax error: unexpected argument ‘080100’
```

**Comment by abhilash1910** [comment]
> @slaren Thanks for the review
> 
> > get_flags is only used with LLAMA_ALL_WARNINGS, yet it also contains optimization flags for the Intel compiler
> 
> @abhilash1910 Could you take a look into this - it does look incorrect to add Intel compiler flags here
> 
> Will wait for @cebtenzzre to review `get_flags` usage and see if we can improve

Yes   it does not appear to be correct as that is in the scope of LLAMA_ALL_WARNINGS. I will remove this. Thanks for pointing this out.

**Comment by ngxson** [comment]
@cebtenzzre Thanks for pointing it out, https://github.com/ggerganov/llama.cpp/pull/5068 was a trial to make the compilation with intel dpcpp yield more performance, but this is pretty much deprecated now after the introduction of sycl. I'm planning to remove these flags.

**Pull request closed without merging**